File: algorithms/644. Maximum Average Subarray II/solution.py
class Solution(object):
    def findMaxAverage(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: float
        """
        # inspired by @ awice
        def avg(i, j):
            return (presum[j+1] - presum[i]) / float(j + 1 - i)

        presum = [0]
        for n in nums:
            presum.append(presum[-1] + n)
        # convex hull processing
        hull = collections.deque()
        result = float('-inf')
        for i in xrange(k - 1, len(nums)):
            while len(hull) > 1 and avg(hull[-2], hull[-1] - 1) >= avg(hull[-2], i - k):
                hull.pop()
            hull.append(i - k + 1)
            while len(hull) > 1 and avg(hull[0], hull[1] - 1) <= avg(hull[0], i):
                hull.popleft()
            result = max(result, avg(hull[0], i))
        return result

# 74 / 74 test cases passed.
# Status: Accepted
# Runtime: 696 ms

File: algorithms/644. Maximum Average Subarray II/solution.cc
class Solution {
public:
    double findMaxAverage(vector<int>& nums, int k) {
        int n = nums.size();
        vector<double> p(n + 1);
        for ( int i = 0; i < n; ++i )
            p[i + 1] = p[i] + nums[i];
        deque<int> h;
        double result = -10000;
        for ( int i = k - 1; i < n; ++i ) {
            while ( h.size() > 1 && avg(p, h[h.size()-2], h.back()-1) >= avg(p, h.back(), i-k))
                h.pop_back();
            h.push_back(i - k + 1);
            while ( h.size() > 1 && avg(p, h[0], i) <= avg(p, h[1], i))
                h.pop_front();
            result = max(result, avg(p, h[0], i));
        }
        return result;
    }

    inline double avg(vector<double> & p, int i, int j) {
        return (p[j+1] - p[i]) / (j + 1 - i);
    }
};

// 74 / 74 test cases passed.
// Status: Accepted
// Runtime: 149 ms
// beats 52.19 %
